Test filter library

lomamech opened the pull request:
- Implement Unit test fror Library Filter
- Update API for patient, Use filters.custom instead
  filters.customMultiParameters

improve(library Filter)  
- Remove filter customMultiParameters

Complete unit test for Filter Library to :
  - test a query that uses both dateFrom() and dateTo()
  - test a query with autoParseStatments
  - test a query with a Uuid is automatically parsed

closes #2313  #3085

lomamech committed:
test(Unit Test Filter Library)
- Implement Unit test fror Library Filter
- Update API for patient, Use filters.custom instead
  filters.customMultiParameters

lomamech committed:
improve(library Filter)
- Remove filter customMultiParameters

lomamech committed:
test(Filter Library)
Complete unit test for Filter Library to :
  - test a query that uses both dateFrom() and dateTo()
  - test a query with autoParseStatments
  - test a query with a Uuid is automatically parsed

jniles commented on server/lib/filter.js:
```diff
@@ -182,18 +182,6 @@ class FilterParser {
     }
   }
 
-  customMultiParameters(filterKey, preparedStatement, preparedValues) {
```
:dancer:

jniles commented on test/server-unit/filter.spec.js:
```diff
@@ -0,0 +1,169 @@
+    filters.fullText('name');
+    const sql = `SELECT t.id, t.name, t.date_object, t.value, t.country_id FROM tables AS t`;
+
+    const formated = filters.applyQuery(sql).trim();
```
Just a small note - "formated" should be spelled "formatted" :)

jniles commented on test/server-unit/filter.spec.js:
```diff
@@ -0,0 +1,169 @@
+describe('filter.js', () => {
+  
+  it('#fullText Format the sql query when filtered by full text', () => {
+    const filters = new filter(object1, { tableAlias : 't' });
```
To make these tests a bit simpler/cleaner, I propose you move all the `new filter(object1, { tableAlias : 't' })` calls into a `beforeEach()` method.  Like this:
```js

let firstFilters;
let firstSql;

beforeEach(() => {
  firstFilters = new Filter(object1, { tableAlias : 't' });
  firstSql = `SELECT t.id, t.name, t.date_object, t.value, t.country_id FROM tables AS t`;
});
```

This will clean up each block of code by focusing only on the test instead of the setup portion.  For example, you first test would then be:
```js
it('#fullText Format the sql query when filtered by full text', () => {

  // fullText matches the SQL "LIKE" filter.
  filters.fullText('name');

  const formatted = filters.applyQuery(sql).trim();
  const expected = `SELECT t.id, t.name, t.date_object, t.value, t.country_id FROM tables AS t WHERE LOWER(t.name) LIKE ?`;

  // assert that the SQL is formatted correctly.
  expect(formatted).to.equal(expected);
});
```

jniles reviewed (commented):
@lomamech, these are really good tests!

One important thing to add to these tests: you should also check that the `filters.parameters()` returns the correct values after a filter is applied.  You already check that the SQL statement is correct - let's just add something like:
```js
const params = filters.parameters();
expect(params).to.deep.equal([1, 2, 3]);
```

This is particularly important when we combine queries, like the `dateFrom()` and `dateTo()` filters ...  if the FilterParser returned them in the wrong order, we wouldn't get the right results!

lomamech committed:
test(Filter Library)
- Clean Up Each block of Code By Comment
- Check that the filters.parameters returns the correct Values after
  filter Applied
